src/vsr/replica.zig

```diff
@@ -2136,15 +2136,40 @@ pub fn ReplicaType(
                     }
                 } else unreachable;
 
-                log.mark.debug("{}: on_do_view_change: lagging primary; forfeiting " ++
-                    "(view={}..{} checkpoint={}..{})", .{
-                    self.replica,
-                    self.view,
-                    next_view,
-                    self.op_checkpoint(),
-                    op_checkpoint_max,
-                });
-                self.transition_to_view_change_status(next_view);
+                // Advancing commit_max for a replica that *could* move to the next checkpoint but
+                // isn't able to because its lagging is crucial for maximizing availability.
+                // Consider the following scenario:
+                // 1. R0 crashes, R1 & R2 form a quorum.
+                // 2. R1 reaches op_checkpoint=59, R2 crashes before it reaches op_checkpoint=59
+                // 3. R0 restarts and state syncs to R1's checkpoint, op_checkpoint=59
+                // 4. R1 crashes and restarts with a corruption in the current checkpoint.
+                //    This corruption restricts R1 & R0 (because it needs to sync those blocks from
+                //    R1) from committing up till commit_max and becoming primary.
+                // 5. R2 restarts but is not able to move to op_checkpoint=59 because it is lagging.
+                // 6. R1, R2 & R3 get stuck in .view_change
+
+                if (self.op_checkpoint_next() == op_checkpoint_max and
+                    self.commit_min < self.op_checkpoint_next_trigger() and
+                    self.op >= self.op_checkpoint_next_trigger())
+                {
+                    const commit_max = @max(
+                        self.commit_max,
+                        DVCQuorum.commit_max(self.do_view_change_from_all_replicas),
+                    );
+
+                    self.commit_max = commit_max;
```

Comment:
The actual problem here seems real enough though! 

I wonder if we could perhaps solve it by relaxing the condition of forfeiting? It seems like we should only forefeit if it is true that **both**:

* cluster is ahead of us by a checkpoint
* the checkponit is confirmed

Reply:
>Hm, I am not sure this would be correct: 

Though the fact that vopr doesn't find any issue here perhaps means that I am wrong about this one?

Reply:
Ah, advancing commit_max without being certain about whether the op we have in our log is correct or not is something that I didn't consider! Great point, thanks a lot for pointing that out.

The correct solution would perhaps involve also invoking `primary_set_log_from_do_view_change_messages`, since we would then keep our headers, head op & commit max up-to-date across view changes... Basically, we're only invoking commit_journal here, when what we *should* be doing is everything (updating log_view, invoking primary_set_log_from_do_view_change_messages and then repair which in turn should perform advance the commit pipeline)?

Reply:
> Though the fact that vopr doesn't find any issue here perhaps means that I am wrong about this one?

This has changed since, we have a failing seed! So I definitely need to think deeper about this change. :)

Also you talk about the checkpoint being confirmed, are you referring to it being durable on a commit-quorum of replicas?

Reply:
> Also you talk about the checkpoint being confirmed, are you referring to it being durable on a commit-quorum of replicas?

Yep! Which I think is “commit max is beyond prepare max for prev checkpoint”

Reply:
Ah, I _think_ I've got it! Thinking back to your earlier comment https://github.com/tigerbeetle/tigerbeetle/pull/2221#discussion_r1717477390, I realized that we require altering _both_ of these solutions to effectively and correctly address this issue... 

Syncing replicas not sending a prepare_ok for any ops past op_prepare_max (if tables haven't yet been synced) ensures that a syncing replica doesn't falsely contribute to the durability of a checkpoint. Specifically, this ensures that no replica would be able to overwrite ops from the previous checkpoint. This dovetails nicely with our discussion in this comment! If a replica is lagging by only a checkpoint and it finds that commit_max < op_prepare_max, it doesn't need to forfeit the view change. It can simpy use WAL repair to catch up since no ops from the lagging replica's checkpoint would have been overwritten yet (since we don't have durability for the *next* checkpoint).

Reply:
Although, I believe the latter change *would* require some changes to [primary_set_log_from_do_view_change_messages](https://github.com/tigerbeetle/tigerbeetle/blob/880d406c08c1a7354a50d536cca2cacfa4e04e1b/src/vsr/replica.zig#L8149), which currently assumes that the header of the canonical DVC is within our op_prepare_max (`assert(header_head.op <= self.op_prepare_max());`) and replaces the replica's head op with the canonical DVC's head op (`self.set_op_and_commit_max(header_head.op, commit_max, @src());`).  Specifically, [primary_set_log_from_do_view_change_messages](https://github.com/tigerbeetle/tigerbeetle/blob/880d406c08c1a7354a50d536cca2cacfa4e04e1b/src/vsr/replica.zig#L8149) would likely have to change the head op to `@min(self.op_prepare_max(), header_head.op)`. Also, we would have to add checks around using the DVC headers to replace our own headers, such that only headers up to op_prepare_max are replaced.

I still haven't thought this through properly though, so this may not be true. 😄

Reply:
Ah yes, quick experiment with running the VOPR after making the above changes *minus* the changes to [primary_set_log_from_do_view_change_messages](https://github.com/tigerbeetle/tigerbeetle/blob/880d406c08c1a7354a50d536cca2cacfa4e04e1b/src/vsr/replica.zig#L8149) that I'd anticipated showed that we would indeed need to make those changes:
```C
[debug] (replica): 1: on_do_view_change: view=25 quorum received
[debug] (replica): 1: on_do_view_change: dvc: replica=0 log_view=24 op=71 commit_min=67 checkpoint=59
[debug] (replica): 1: on_do_view_change: dvc: header: replica=0 op=71 checksum=208545351406523309574206797536401977373 nack=false present=true type=valid
[debug] (replica): 1: on_do_view_change: dvc: header: replica=0 op=70 checksum=60171286467781141794056692787268822268 nack=false present=true type=valid
[debug] (replica): 1: on_do_view_change: dvc: header: replica=0 op=69 checksum=168193058532343774027948307381785896824 nack=false present=true type=valid
[debug] (replica): 1: on_do_view_change: dvc: header: replica=0 op=68 checksum=223364609765652245498886553981331736644 nack=false present=true type=valid
[debug] (replica): 1: on_do_view_change: dvc: header: replica=0 op=67 checksum=299943675278850709487137762353369328066 nack=false present=true type=valid
[debug] (replica): 1: on_do_view_change: dvc: replica=1 log_view=23 op=67 commit_min=62 checkpoint=39
[debug] (replica): 1: on_do_view_change: dvc: header: replica=1 op=67 checksum=299943675278850709487137762353369328066 nack=false present=true type=valid
[debug] (replica): 1: on_do_view_change: dvc: header: replica=1 op=66 checksum=325996494377482587983529677958291707278 nack=false present=true type=valid
thread 19040384 panic: reached unreachable code
/Users/chaitanyabhandari/Desktop/tigerbeetle/tigerbeetle/zig/lib/std/debug.zig:412:14: 0x100ba170f in assert (vopr)
    if (!ok) unreachable; // assertion failure
             ^
/Users/chaitanyabhandari/Desktop/tigerbeetle/tigerbeetle/src/vsr/replica.zig:8207:19: 0x100ca0a43 in primary_set_log_from_do_view_change_messages (vopr)
            assert(header_head.op <= self.op_prepare_max());
                  ^
```

Reply:
> which currently assumes that the header of the canonical DVC is within our op_prepare_max (assert(header_head.op <= self.op_prepare_max());)

Ohh, this is fascinating! I’ll double check this tomorrow, but so far it indeed looks like it could be the case that, during view change, there is a checkpoint ahead of yours, and even the head op comes from this next checkpoint, and yet you might need to become a primary, because the checkpoint is borked, and only wal repair would work.

One question though: is this forfeiting change actually repeated to “not sending prepare ok”s one? It looks like these might be two unrelated issues? If that is the case, than perhaps we  can split this into two prs.

Intuitively it seems that “don’t falsely confirm checkpoint as durable if you are still state syncing” is much easier than “don’t forfeit if checkpoint might not be durable”.

Reply:
Agreed, these are two separate issues and makes sense to tackle them separately!

Also, there's a failing liveness issue on main right now. I will investigate it tomorrow, hopefully it is related to the latter issue, i.e. “don’t forfeit if checkpoint might not be durable”. 🤞🏼